Add Engine Switching

alexsporn opened the pull request.

karimodm committed:
Serialize & deserialize attestations

karimodm committed:
WIP

karimodm committed:
Deal with commitment fork events

alexsporn committed:
Prepared engine switching

alexsporn committed:
Calculate weight of forking point with received attestations

alexsporn committed:
Merge branch 'develop' of github.com:iotaledger/goshimmer into feat/chain-switching

# Conflicts:
#	go.mod
#	go.sum

alexsporn committed:
Use correct hive.go

alexsporn committed:
Handle attestation requests with a range of epochs

alexsporn committed:
Do not create a new chain manager when switching engines

alexsporn committed:
Sort the attestations to apply them in order

alexsporn committed:
Checked all access to the engine to be protected by the active engine mutex

alexsporn committed:
Reverted attestations being stored by public key

Added helpers to attestation to get the issuer id and to verify the signature

alexsporn committed:
Updated the docker network feature snapshot

alexsporn committed:
Removed temporary code

alexsporn committed:
Evict the chain forking events

alexsporn committed:
Don’t rely on the peer sending the correct epochs

alexsporn committed:
go mod tidy on integration tester

alexsporn committed:
Fixed linter and compiler issues

alexsporn committed:
Fixed tests that broke due to the change of signature in the snapshotcreator

alexsporn committed:
Fixed chainmanager tests

alexsporn committed:
Fixed linter issues

alexsporn committed:
Refactored part of the engine forking code into a new EngineManager

alexsporn committed:
Linter fixes

alexsporn committed:
Remove sync.Once from module provider

alexsporn committed:
Add TODOs

alexsporn committed:
Add EngineManager tests

alexsporn committed:
Merge branch 'develop' of github.com:iotaledger/goshimmer into feat/chain-switching

alexsporn committed:
Started working on engine switching tests with multiple virtual nodes

alexsporn committed:
Expose ChainManager events

alexsporn committed:
Fixed main partition not exisiting in the MockedNetwork

alexsporn committed:
Fixed partitioned networks not accepting and commiting

alexsporn committed:
Fixed blocks not committing to latest epoch in the mocked network

alexsporn committed:
Added events to protocol about engine switching

alexsporn committed:
Expose engine network events

alexsporn committed:
register node aliases in test

alexsporn committed:
Added assertion that the nodes have the same chain

alexsporn committed:
assign the funds to the node identities when creating a snapshot

add optional engine options to be used when creating a snapshot

alexsporn committed:
added the forking point to the attestations request

alexsporn committed:
Add a way to the the latest commitment of a given chain

alexsporn committed:
Use the forking point to request attestations

alexsporn committed:
set the correct chain id after forking the engine

alexsporn committed:
Attach the candidate engine blockrequester to the network

alexsporn committed:
Add a Sign function to the models that sets the issuer and signs the block with a given keypair

alexsporn committed:
updated mocked node code to properly sign transactions

logging not also attaches to both the main and the candidate engine

alexsporn committed:
Added some block activity to let the candidate engine sync

alexsporn committed:
Wait until nodes switched the engine instead of sleeping for a hardcoded time

alexsporn committed:
Also send blockIDs with the attestations so that the candidate engine knows which blocks to request

alexsporn committed:
Change the order inside the snapshot so that the eviction state is properly initialized after importing and rolling back the ledger

alexsporn committed:
Add BlockProcessed event to the engine

alexsporn committed:
Remove wait on storage shutdown and added TODO

alexsporn committed:
Updated engine switching test

alexsporn committed:
Wait until the engine workerpools are shutdown before shutting down the engine storage

alexsporn committed:
Give the engines a name for the logs

alexsporn committed:
Properly handle the chainmanager commitment publishing

alexsporn committed:
check if the engine is stopped in the sybilprotection

alexsporn committed:
Don’t expect all nodes to commit to the same epoch in the test, but compare they all commited after merging and that the commits are equal

alexsporn committed:
Avoid crash if switchEngines is called multiple times

alexsporn committed:
Reset the visited identities check at each epoch

alexsporn committed:
Better logging

alexsporn committed:
when we receive a commitment request, do not look it up in the ChainManager but in the storage, else we might answer with commitments we did not issue ourselves and for which we cannot provide attestations

alexsporn committed:
Fixed engine manager tests

alexsporn committed:
Fixed crash in hive.go

alexsporn committed:
Fixed chainmanager tests

alexsporn committed:
Fixed multiple race conditions

alexsporn committed:
Disable commitment byte caching to avoid a data race in immutable

alexsporn committed:
Fixed comparison in test

hmoog commented on packages/core/commitment/commitment.go:
```diff
@@ -30,11 +30,11 @@ func New(index epoch.Index, prevID ID, rootsID types.Identifier, cumulativeWeigh
 		RootsID:          rootsID,
 		CumulativeWeight: cumulativeWeight,
-	})
+	}, false) //TODO: fix the data-race in immutable so that we can enable byte caching
```
don't forget to revert this

hmoog commented on packages/core/commitment/commitment.go:
```diff
@@ -30,11 +30,11 @@ func New(index epoch.Index, prevID ID, rootsID types.Identifier, cumulativeWeigh
 
 func NewEmptyCommitment() (newCommitment *Commitment) {
-	return model.NewImmutable[Commitment](&commitment{})
+	return model.NewImmutable[Commitment](&commitment{}, false)
```
don't forget to revert this

hmoog commented on packages/network/manualpeering/manualpeering.go:
```diff
@@ -366,7 +367,7 @@ func (m *Manager) changeNeighborStatus(neighbor *p2p.Neighbor, connStatus Connec
 
 func (m *Manager) connectionDirection(peerPK ed25519.PublicKey) (ConnectionDirection, error) {
-	result := bytes.Compare(m.local.PublicKey().Bytes(), peerPK.Bytes())
+	result := bytes.Compare(lo.PanicOnErr(m.local.PublicKey().Bytes()), lo.PanicOnErr(peerPK.Bytes()))
```
since we already return an error in the method, we should maybe not use lo.PanicOnError

hmoog committed:
Feat: cleaned up some code

hmoog commented on packages/protocol/chainmanager/manager.go:
```diff
@@ -103,6 +132,21 @@ func (m *Manager) registerCommitment(commitment *commitment.Commitment) (isSolid
 
+	chainCommitment, created := m.Commitment(commitment.ID(), true)
+
+	if !created {
```
I think this should be behind the PublishCommitment further down the line and only emitted if we store the block

hmoog commented on packages/protocol/engine/clock/timeupdate.go:
```diff
@@ -34,6 +34,7 @@ func (c *timeUpdate) Update(now, newTime time.Time) (updated bool) {
 	// the local wall clock should never be before the accepted Time unless we are eclipsed by malicious actors or our
 	// own Time is clearly in the past
 	if now.Before(newTime) {
+		//TODO: a block with a future time panics this
```
We most probably should have a filter that doesn't allow to process future blocks, so they can also not get solid / confirmed

hmoog commented on packages/protocol/engine/consensus/epochgadget/gadget.go:
```diff
@@ -43,6 +43,14 @@ func (g *Gadget) LastConfirmedEpoch() epoch.Index {
+	defer g.Unlock()
+
+	g.lastConfirmedEpoch = i
+	g.Events.EpochConfirmed.Trigger(i)
```
the trigger should maybe happen outside the scope of the lock to prevent possible deadlocks

hmoog commented on packages/protocol/tipmanager/tipmanager.go:
```diff
@@ -78,12 +79,24 @@ func New(schedulerBlockRetrieverFunc blockRetrieverFunc, opts ...options.Option[
 }
 
 func (t *TipManager) AddTip(block *scheduler.Block) {
+	t.tipsMutex.RLock()
```
Is a readlock correct here?

hmoog commented on tools/integration-tests/tester/framework/network.go:
```diff
@@ -401,7 +403,7 @@ func (n *Network) createPeers(ctx context.Context, numPeers int, networkConfig C
 		conf.AutoPeering.Enabled = true
 		conf.AutoPeering.EntryNodes = []string{
-			fmt.Sprintf("%s@%s:%d", base58.Encode(n.entryNode.Identity.PublicKey().Bytes()), n.entryNode.Name(), peeringPort),
+			fmt.Sprintf("%s@%s:%d", base58.Encode(lo.PanicOnErr(n.entryNode.Identity.PublicKey().Bytes())), n.entryNode.Name(), peeringPort),
```
return error in methid instead of lo.PanicOnError

hmoog reviewed (changes requested):
A few comments ...

alexsporn committed:
Use new hive.go with data-race fix

alexsporn committed:
Revert "Disable commitment byte caching to avoid a data race in immutable"

This reverts commit 050f6cc11a30985c3606438ffe6a47b2a8303de7.

alexsporn committed:
Revert "Fixed comparison in test"

This reverts commit a8a404888ef29d6c7fe13f85fdc1ea9faf4a45ea.

alexsporn committed:
Fix review comments

alexsporn committed:
Use the forking point against the main chain when triggering a fork detected event

alexsporn committed:
Use workerpools for all event attachments inside the engine so that we are sure we can shut it down properly

alexsporn committed:
Reorg the main chain inside the chain manager after switching engines

alexsporn committed:
Fix nil pointer dereference in chain manager

alexsporn committed:
Always register children if we are processing our own commitment

alexsporn committed:
Fixed common synchronization integration test by issuing more blocks and witht delay, so that the new node has time to commit epochs

alexsporn committed:
Propagate solidity and reorg the main-chain when issuing own commitments in the chain manager

alexsporn committed:
Wrap errors in the chain manager and handle errors when re-organizing the main chain

alexsporn committed:
Merge branch 'develop' of github.com:iotaledger/goshimmer into feat/chain-switching

alexsporn committed:
Updated hive.go

alexsporn committed:
Pass correct VM to the integration test snapshot creator

alexsporn committed:
Link to the network events correctly

alexsporn committed:
Remove chain extension check

alexsporn committed:
Added missing error handling

alexsporn committed:
Add missing mutex locking in chain manager

alexsporn committed:
Fixed consensus double spend integration test

alexsporn committed:
Fixed the future tips test

alexsporn commented on plugins/peer/plugin.go:
```diff
@@ -134,8 +134,13 @@
+	}
 	prvKeyCfg := ed25519.PrivateKeyFromSeed(cfgSeed)
-	if !bytes.Equal(prvKeyCfg.Bytes(), prvKeyDB.Bytes()) {
+	prvKeyCfgBytes, err := prvKeyCfg.Bytes()
```
Already fixed

alexsporn committed:
Made the lifecycle event traits accept multiple callbacks

alexsporn committed:
Add workerpools to protocol

Add WaitWorkerPoolsEmpty to protocol and engine

alexsporn committed:
Fixed tipmanager test

alexsporn committed:
Detach all closures attached by the protocol when Shutdown is called

alexsporn committed:
Added Runnable trait

alexsporn committed:
Made protocol and all sub-components implement Runnable trait

alexsporn committed:
Added workers to the whole stack

Refactored tests to avoid composition of the test frameworks

hmoog committed:
Refactor: refactored SnapshotCreator API

hmoog committed:
Feat: made vm mandatory param

alexsporn committed:
Disabled leak detection in ledger objectstorage

alexsporn committed:
Removed debug printf

alexsporn committed:
Allow the creation of snapshots with the MockVM

Added balance to the MockVM outputs so that we can initialize the mana vectors from the snapshot

alexsporn committed:
Fixed ratesetter tests

alexsporn committed:
fixed retainer tests

alexsporn committed:
Fixed engine stop and resume test

alexsporn committed:
use correct vm in engine transactions forward and rollback and added asserts

alexsporn committed:
Fixed protocol engine shutdown and restart with permanent storage

alexsporn committed:
fixed scheduler shutdown hanging if it was never started

alexsporn committed:
hook inside the weightedset instead of attach

alexsporn committed:
cleaned up imports

alexsporn committed:
Fixed engine switching tests

alexsporn committed:
Fixed unit tests

alexsporn committed:
Fixed integration test snapshot creation

alexsporn committed:
linter and test fixes

alexsporn committed:
Updated hive.go with latest fixes

alexsporn committed:
Fixed import

alexsporn committed:
Increased timeouts for the tests on the CI

alexsporn committed:
Fixed blockissuer

alexsporn committed:
Increase timeouts in engine switching test

alexsporn committed:
Updated docker feature network snapshot

alexsporn committed:
process commitments generated by candidate engine

alexsporn committed:
Unregister network protocol when protocol is shutdown

alexsporn committed:
simplify handling of requested blocks when using multiple engines

alexsporn committed:
Added print in engine test

alexsporn committed:
wait on all nodes when issuing the initial manually woven tangle

jonastheis commented on tools/genesis-snapshot/main.go:
```diff
@@ -143,7 +155,9 @@ func init() {
-	e := engine.New(s, dpos.NewProvider(), mana1.NewProvider())
+	defer s.Shutdown()
+
+	e := engine.New(workerpool.NewGroup("Diagnostics"), s, dpos.NewProvider(), mana1.NewProvider())
```
add `defer e.Shutdown()`?

jonastheis commented on tools/genesis-snapshot/main.go:
```diff
@@ -93,7 +97,7 @@ func main() {
 	initialAttestationsSlice := createInitialAttestations()
 
-	snapshotcreator.CreateSnapshot(protocol.DatabaseVersion, snapshotFileName, genesisTokenAmount, genesisSeed, manaDistribution, initialAttestationsSlice)
+	snapshotcreator.CreateSnapshot(workerpool.NewGroup("CreateSnapshot"), protocol.DatabaseVersion, snapshotFileName, genesisTokenAmount, genesisSeed, manaDistribution, initialAttestationsSlice, new(devnetvm.VM))
```
Having this call and passing in a workerpool group on first sight seems to be a non-blocking call. Can we remove the workerpool parameter here as it seems to be not really used (except to instantiate an engine) within this function?

jonastheis commented on packages/core/snapshotcreator/snapshotcreator.go:
```diff
@@ -42,26 +46,32 @@ func CreateSnapshot(databaseVersion database.Version, snapshotFileName string, g
 		// pledge to ID but send funds to random address
-		output, outputMetadata = createOutput(devnetvm.NewED25519Address(ed25519.GenerateKeyPair().PublicKey), value, nodeID, 0)
+		nodeID := identity.NewID(nodePublicKey)
+		output, outputMetadata := createOutput(ledgerVM, nodePublicKey, value, nodeID, 0)
```
comment is not correct anymore. the funds are not send to a random address/burned. also the comment of the function is wrong.

alexsporn committed:
Updated docker-network snapshot

jonastheis commented on packages/protocol/engine/tangle/booker/booker.go:
```diff
@@ -423,27 +429,25 @@ func (b *Booker) strongChildren(block *Block) []*Block {
 			}
 		}
-	}))
+	}, b.Workers.CreatePool("Booker"))
```
This means we create a new unbounded workerpool with `2*runtime.NumCPU()` right? Isn't that too many idle goroutines for "just" handling the `TransactionBooked` event?

jonastheis commented on packages/protocol/engine/tangle/booker/testframework.go:
```diff
@@ -99,38 +103,41 @@ func (t *TestFramework) AssertBlockConflictsUpdateCount(blockConflictsUpdateCoun
 
 func (t *TestFramework) setupEvents() {
-	t.Booker.Events.BlockBooked.Hook(event.NewClosure(func(metadata *Block) {
+	t.Instance.Events.BlockBooked.Hook(event.NewClosure(func(metadata *Block) {
```
Why not the new style to hook to events?

jonastheis commented on packages/protocol/engine/tangle/testframework.go:
```diff
@@ -5,127 +5,64 @@ import (
-		if t.optsLedger != nil {
-			panic("using the TestFramework with Ledger and LedgerOptions simultaneously is not allowed")
-		}
+func NewDefaultTestFramework(t *testing.T, workers *workerpool.Group, optsTangle ...options.Option[Tangle]) *TestFramework {
```
Having different constructors and combining them together is much easier to read and much less prone to errors due to unexpected behavior because of some default option parameters. I like!

jonastheis commented on packages/protocol/engine/throughputquota/mana1/throughputquota.go:
```diff
@@ -179,7 +183,8 @@ func (m *ThroughputQuota) init() {
 	m.TriggerInitialized()
 
-	m.engine.Ledger.Events.TransactionAccepted.Attach(event.NewClosure(func(event *ledger.TransactionEvent) {
+	wp := m.workers.CreatePool("ThroughputQuota")
```
Do we need `2*runtime.NumCPU()` workers? I'd suggest to validate all calls to `CreatePool` in this regard

jonastheis commented on packages/protocol/engine/consensus/blockgadget/gadget.go:
```diff
@@ -233,15 +239,15 @@ func (a *Gadget) EvictUntil(index epoch.Index) {
-	a.tangle.VirtualVoting.Events.SequenceTracker.VotersUpdated.Attach(event.NewClosure(func(evt *sequencetracker.VoterUpdatedEvent) {
-		a.RefreshSequence(evt.SequenceID, evt.NewMaxSupportedIndex, evt.PrevMaxSupportedIndex)
-	}))
+	wp := a.workers.CreatePool("Gadget")
```
Is it necessary to have in total 3 worker pools with `2*runtime.NumCPU()` each for the block gadget?

jonastheis commented on packages/protocol/engine/consensus/blockgadget/gadget.go:
```diff
@@ -314,9 +320,12 @@ func (a *Gadget) markAsAccepted(block *Block) (err error) {
 	if block.SetAccepted() {
+		if block.ID().String() == "BlockID(Block5)" {
+			fmt.Println("foo")
+		}
```
Remove debug print

jonastheis commented on packages/protocol/engine/consensus/blockgadget/testframework.go:
```diff
@@ -11,91 +11,84 @@ import (
 			t.test.Logf("ACCEPTED: %s", metadata.ID())
+			if metadata.ID().String() == "BlockID(Block5)" {
+				_ = ""
+			}
```
remove debug print

jonastheis commented on packages/protocol/engine/consensus/epochgadget/gadget.go:
```diff
@@ -43,27 +47,33 @@ func (g *Gadget) LastConfirmedEpoch() epoch.Index {
+	event.AttachWithWorkerPool(g.tangle.VirtualVoting.Events.EpochTracker.VotersUpdated, func(evt *epochtracker.VoterUpdatedEvent) {
 		g.refreshEpochConfirmation(evt.PrevLatestEpochIndex, evt.NewLatestEpochIndex)
-	}))
+	}, g.workers.CreatePool("Refresh"))
```
`2*runtime.NumCPU()` necessary?

jonastheis commented on packages/protocol/ledger/ledger.go:
```diff
@@ -189,10 +190,8 @@ func (l *Ledger) processTransaction(tx utxo.Transaction) (err error) {
 func (l *Ledger) processConsumingTransactions(outputIDs utxo.OutputIDs) {
 	for it := l.Utils.UnprocessedConsumingTransactions(outputIDs).Iterator(); it.HasNext(); {
 		txID := it.Next()
-		event.Loop.Submit(func() {
```
Is this correct in the data flow? I.e. coming from the Tangle we book a tx and don't want to continue booking `UnprocessedConsumingTransactions` in the same goroutine but instead return to the tangle and continue booking our block. I think it should be queued in `l.wokerPool`

alexsporn replied:
We are already running in a workerpool inside the ledger

jonastheis commented on packages/protocol/tipmanager/tipmanager.go:
```diff
@@ -78,12 +79,24 @@ func New(schedulerBlockRetrieverFunc blockRetrieverFunc, opts ...options.Option[
 
 func (t *TipManager) AddTip(block *scheduler.Block) {
+	t.tipsMutex.Lock()
+	defer t.tipsMutex.Unlock()
```
What is this lock for exactly? Seems like it's protecting the instances, which means it could always be a read lock and is probably a superset of the `evictionMutex`. Can these be combined?

jonastheis commented on packages/storage/utils/directory.go:
```diff
@@ -29,4 +30,36 @@ func (d *Directory) Path(relativePathElements ...string) (path string) {
+	var dirs []string
+	basePath := filepath.Base(d.path)
+
+	err := filepath.WalkDir(d.path, func(path string, dirEntry os.DirEntry, err error) error {
```
Seems to be a bit complicated for what it is doing. `os.ReadDir()` as described [here](https://stackoverflow.com/questions/14668850/list-directory-in-go) should be simpler and probably faster

jonastheis commented on packages/protocol/tipmanager/testframework.go:
```diff
@@ -107,34 +104,34 @@ func (t *TestFramework) setupEvents() {
-	t.mockAcceptance.BlockAcceptedEvent.Attach(event.NewClosure(func(block *blockgadget.Block) {
-		require.NoError(t.test, t.engine.NotarizationManager.NotarizeAcceptedBlock(block.ModelsBlock))
-	}))
+	event.Hook(t.mockAcceptance.BlockAcceptedEvent, func(block *blockgadget.Block) {
```
Is `event.Hook(` here expected? As it was `Attach` before.

alexsporn replied:
It is easier to hook inside the tests

jonastheis commented on packages/protocol/tipmanager/testframework.go:
```diff
@@ -107,34 +104,34 @@ func (t *TestFramework) setupEvents() {
-	t.engine.NotarizationManager.Events.EpochCommitted.Attach(event.NewClosure(func(details *notarization.EpochCommittedDetails) {
-		t.TipManager.PromoteFutureTips(details.Commitment)
-	}))
+	event.Hook(t.Engine.NotarizationManager.Events.EpochCommitted, func(details *notarization.EpochCommittedDetails) {
```
Is `event.Hook(` here expected? As it was `Attach` before.

alexsporn replied:
It is easier to hook inside the tests

jonastheis commented on packages/protocol/tipmanager/testframework.go:
```diff
@@ -107,34 +104,34 @@ func (t *TestFramework) setupEvents() {
-	t.engine.EvictionState.Events.EpochEvicted.Attach(event.NewClosure(func(index epoch.Index) {
-		t.TipManager.Evict(index)
-	}))
+	event.Hook(t.Engine.EvictionState.Events.EpochEvicted, func(index epoch.Index) {
```
Is `event.Hook(` here expected? As it was `Attach` before.

alexsporn replied:
It is easier to hook inside the tests

jonastheis commented on packages/protocol/engine/sybilprotection/dpos/sybilprotection.go:
```diff
@@ -68,12 +72,14 @@ func NewSybilProtection(engineInstance *engine.Engine, opts ...options.Option[Sy
+				event.AttachWithWorkerPool(s.engine.Events.Tangle.BlockDAG.BlockSolid, func(block *blockdag.Block) {
 					s.markValidatorActive(block.IssuerID(), block.IssuingTime())
-				}))
+				}, s.workers.CreatePool("SybilProtection"))
```
`2*runtime.NumCPU()` workers?

jonastheis commented on packages/protocol/engine/notarization/attestation.go:
```diff
@@ -62,3 +63,20 @@ func (a Attestation) Bytes() (bytes []byte, err error) {
 }
+
+func (a *Attestation) IssuerID() identity.ID {
+	return identity.NewID(a.IssuerPublicKey)
```
Seems like this computed value is used often. Maybe we can just create a non-serialized property on object creation to avoid calling this over and over again?

alexsporn replied:
This does not seem to be called so often

jonastheis commented on packages/protocol/engine/engine.go:
```diff
@@ -280,176 +276,174 @@ func (e *Engine) initLedger() {
 			e.Events.Error.Trigger(errors.Wrapf(err, "failed to attach block with %s (issuerID: %s)", block.ID(), block.IssuerID()))
 		}
-	}))
+	}, e.Workers.CreatePool("Tangle.Attach", 1))
```
Sure we only want 1 worker here?

alexsporn committed:
Removed debug prints

alexsporn committed:
Using new event.Hook syntax inside all the packages

alexsporn committed:
review fixes

alexsporn committed:
Combined both mutexes inside the TipManager

alexsporn committed:
Removed workers parameter from CreateSnapshot

jonastheis commented on packages/protocol/chainmanager/manager.go:
```diff
@@ -3,24 +3,36 @@ package chainmanager
 	SnapshotCommitment  *Commitment
 	CommitmentRequester *eventticker.EventTicker[commitment.ID]
 
+	//TODO: change to memstorage.EpochStorage?
 	commitmentsByID      map[commitment.ID]*Commitment
```
This will grow forever no?

alexsporn replied:
Yes, that is why I added that TODO. We need to decide what should happen here

jonastheis commented on packages/protocol/chainmanager/manager.go:
```diff
@@ -40,32 +53,27 @@ func NewManager(snapshot *commitment.Commitment) (manager *Manager) {
-	manager.Events.CommitmentMissing.Attach(event.NewClosure(manager.CommitmentRequester.StartTicker))
-	manager.Events.MissingCommitmentReceived.Attach(event.NewClosure(manager.CommitmentRequester.StopTicker))
+	manager.Events.CommitmentMissing.Hook(event.NewClosure(manager.CommitmentRequester.StartTicker))
+	manager.Events.MissingCommitmentReceived.Hook(event.NewClosure(manager.CommitmentRequester.StopTicker))
```
Use new style for events

alexsporn committed:
Reduced the amount of workers in each workerpool

jonastheis commented on packages/protocol/chainmanager/manager.go:
```diff
@@ -77,23 +85,74 @@ func (m *Manager) ProcessCommitment(commitment *commitment.Commitment) (isSolid
 	}
 
+	// The forking point should be at least 3 epochs in the past w.r.t this commitment index.
+	//TODO: replace 3 with values depending on the configuration
+	latestChainCommitment := chain.LatestCommitment()
+	if latestChainCommitment.ID().Index()-forkingPoint.ID().Index() < 3 {
```
Fix TODO and make it configurable

jonastheis commented on packages/protocol/chainmanager/manager.go:
```diff
@@ -3,24 +3,36 @@ package chainmanager
 	commitmentsByID      map[commitment.ID]*Commitment
 	commitmentsByIDMutex sync.Mutex
 
+	commitmentsByForkingPoint *memstorage.EpochStorage[commitment.ID, *ForkDetectedEvent]
```
Seems to be a bit odd to store an event here. Maybe we can just rename this struct to something else?

jonastheis commented on packages/protocol/chainmanager/manager.go:
```diff
@@ -77,23 +85,74 @@ func (m *Manager) ProcessCommitment(commitment *commitment.Commitment) (isSolid
+}
+
+func (m *Manager) registerCommitment(commitment *commitment.Commitment) (isNew bool, isSolid bool, wasForked bool, chainCommitment *Commitment) {
+	m.commitmentEntityMutex.Lock(commitment.ID())
+	defer m.commitmentEntityMutex.Unlock(commitment.ID())
```
Can this lead to a deadlock if another goroutine is currently in `processCommitment` to propagate solidity to the children. There we hold the lock of the parent in `processCommitment` and update the child in `propagateSolidity` where we lock the child. Here we do it the opposite way.

alexsporn committed:
Added new debug config parameters to allow the node to panic on fork detection

jonastheis commented on packages/protocol/chainmanager/manager.go:
```diff
@@ -148,6 +217,105 @@ func (m *Manager) Commitments(id commitment.ID, amount int) (commitments []*Comm
+
+func (m *Manager) Evict(index epoch.Index) {
+	//TODO: do we need to evict more stuff?
+	m.commitmentsByForkingPoint.Evict(index)
```
We don't need an eviction lock here?

What about `commitmentsByID`?

alexsporn replied:
No need for a lock, because the epochStorage and the Storage have locks already.
Regarding the `commitmentsByID` we need to discuss. I think the idea was to keep them all in memory, but maybe @karimodm and @hmoog can chime in

jonastheis commented on packages/protocol/testframework.go:
```diff
@@ -31,59 +38,47 @@ const genesisTokenAmount = 100
-	AcceptanceTestFramework = blockgadget.TestFramework
-)
-
 type EngineTestFramework struct {
```
Shouldn't this live in `engine/` ?

alexsporn committed:
Added missing mutex lock in weightedset detected by race detector

alexsporn committed:
Move the engine testframework to the engine package

alexsporn committed:
Renamed ForkDetectedEvent to Fork

alexsporn committed:
Increased the Tangle.Attach workerpool to 2

alexsporn committed:
Only switch to candidate engine if it has at least the same latest commitment index as the main engine

alexsporn committed:
shutdown old candidate engines if new ones are created

alexsporn committed:
Made the minimum fork depth for detection configurable

Fixed fork re-detection after eviction

alexsporn committed:
Fill up the chain manager from storage if we restart the node instead of asking peers for missing commitments

jonastheis commented on packages/protocol/chainmanager/manager.go:
```diff
@@ -148,6 +219,109 @@ func (m *Manager) Commitments(id commitment.ID, amount int) (commitments []*Comm
+	//TODO: do we need to evict more stuff?
+
+	// Forget about the forks that we detected at that epoch so that we can detect them again if they happen
+	evictedForkDetections := m.forkingPointsByCommitments.Evict(index)
```
I still think this eviction process as a whole should be atomic (by introducing an evicitionMutex like in other components). What exactly happens if there's reads while this stuff is half evicted?

jonastheis commented on packages/protocol/chainmanager/manager_test.go:
```diff
@@ -94,24 +104,159 @@ func TestManager(t *testing.T) {
+			lo.Return1(tf.Instance.Commitment(tf.EC("3"))),
+			lo.Return1(tf.Instance.Commitment(tf.EC("2"))),
+			lo.Return1(tf.Instance.Commitment(tf.EC("1"))),
+			lo.Return1(tf.Instance.Commitment(tf.EC("Genesis"))),
```
Maybe put this into the testframework so it's reusable? Same stuff in line 225

jonastheis commented on packages/protocol/enginemanager/enginemanager.go:
```diff
@@ -0,0 +1,179 @@
+	return instance, nil
+}
+
+func (e *EngineInstance) InitializeWithSnapshot(snapshotPath string) error {
```
Separate `EngineInstance` and `EngineManager` at least with a region to make it more readable

jonastheis commented on packages/protocol/enginemanager/enginemanager.go:
```diff
@@ -0,0 +1,179 @@
+func (m *EngineManager) ForkEngineAtEpoch(index epoch.Index) (*EngineInstance, error) {
+	// Dump a snapshot at the target index
+	snapshotPath := filepath.Join(os.TempDir(), fmt.Sprintf("snapshot_%d_%s.bin", index, lo.PanicOnErr(uuid.NewUUID()).String()))
+	if err := m.activeInstance.Engine.WriteSnapshot(snapshotPath, index); err != nil {
```
How expensive is this `WriteSnapshot`? Does it need to write the entire ledgerstate or is it some sort of incremental diff that is written? What if it was 1TB big?

alexsporn replied:
Yes, I had the same comment to @karimodm, but right now that is the only way we have to "clone" the state of an engine. It is good enough for the MVP but for production we need a way to stream it from one database to the other.

jonastheis commented on packages/protocol/enginemanager/enginemanager_test.go:
```diff
@@ -0,0 +1,189 @@
+	// Block in epoch 11
+	tf.BlockDAG.CreateBlock("11.A", models.WithStrongParents(tf.BlockDAG.BlockIDs("2.D")), models.WithIssuer(identitiesMap["A"]))
+	tf.BlockDAG.IssueBlocks("11.A")
+	workers.Wait()
```
Should not be necessary as it waits in `tf.BlockDAG.IssueBlocks`

jonastheis commented on packages/protocol/enginemanager/enginemanager_test.go:
```diff
@@ -0,0 +1,189 @@
+	// Time has advanced to epoch 10 because of A.5, rendering 10 - MinimumCommittableAge(6) = 4 epoch committable
+	require.Eventually(t, func() bool {
+		return tf.Instance.Storage.Settings.LatestCommitment().Index() == epoch.Index(4)
+	}, time.Second, 100*time.Millisecond)
```
1 second wait time might be too short?

alexsporn replied:
1 second is enough here,, we only need to wait for the issued blocks to be accepted and that will trigger an epoch commitment

jonastheis commented on packages/protocol/protocol.go:
```diff
@@ -45,268 +43,520 @@ type Protocol struct {
-	p.networkProtocol.Events.EpochCommitmentRequestReceived.Attach(event.NewClosure(func(event *network.EpochCommitmentRequestReceivedEvent) {
-		if requestedCommitment, _ := p.chainManager.Commitment(event.CommitmentID); requestedCommitment != nil && requestedCommitment.Commitment() != nil {
-			p.networkProtocol.SendEpochCommitment(requestedCommitment.Commitment(), event.Source)
+	wpCommitments := p.Workers.CreatePool("NetworkEvents.EpochCommitments") // Using just 1 worker to avoid contention
```
```suggestion
	wpCommitments := p.Workers.CreatePool("NetworkEvents.EpochCommitments", 1) // Using just 1 worker to avoid contention
```

jonastheis commented on packages/protocol/protocol.go:
```diff
@@ -45,268 +43,520 @@ type Protocol struct {
-	p.engine = engine.New(p.storage, p.optsSybilProtectionProvider, p.optsThroughputQuotaProvider, p.optsEngineOptions...)
+	}, wpCommitments)
+
+	wpAttestations := p.Workers.CreatePool("NetworkEvents.Attestations") // Using just 1 worker to avoid contention
```
```suggestion
	wpAttestations := p.Workers.CreatePool("NetworkEvents.Attestations", 1) // Using just 1 worker to avoid contention
```

jonastheis commented on packages/protocol/protocol.go:
```diff
@@ -45,268 +43,520 @@ type Protocol struct {
-	p.activeEngineMutex.RLock()
-	defer p.activeEngineMutex.RUnlock()
+func (p *Protocol) ProcessAttestations(forkingPoint *commitment.Commitment, blockIDs models.BlockIDs, attestations *orderedmap.OrderedMap[epoch.Index, *set.AdvancedSet[*notarization.Attestation]], source identity.ID) {
+	fmt.Println("Received attestations for", forkingPoint.ID())
```
remove debug print

jonastheis commented on packages/protocol/protocol.go:
```diff
@@ -45,268 +43,520 @@ type Protocol struct {
+
+	var calculatedCumulativeWeight int64
+	mainEngine.Engine.NotarizationManager.PerformLocked(func(m *notarization.Manager) {
+		// Calculate the difference between the latest commitment ledger and the ledger at the snapshot target index
```
Couldn't we calculate the weight diff for each epoch as part of the commitments and store it until it can be pruned. Then we only need to sum up those diffs here for the epochs instead of every time streaming through all the spent/created outputs. Or is it possible to get this information from the attestations of our current engine themselves?

alexsporn replied:
This is what was possible with the current available means. This is not called very often, so I don't see a problem in streaming it here. We should think about this whenever we tackle the ledger refactoring though.

jonastheis commented on packages/protocol/protocol.go:
```diff
@@ -45,268 +43,520 @@ type Protocol struct {
+	if err := candidateEngine.Engine.Storage.Settings.SetChainID(forkedEvent.ForkingPoint.ID()); err != nil {
+		p.Events.Error.Trigger(errors.Wrap(err, "error setting the ChainID on the forked engine"))
+		candidateEngine.Shutdown()
+		_ = candidateEngine.RemoveFromFilesystem()
```
Would be good to still log this error

jonastheis commented on packages/protocol/protocol.go:
```diff
@@ -45,268 +43,520 @@ type Protocol struct {
+
+	p.Events.MainEngineSwitched.Trigger(p.MainEngineInstance())
+
+	// TODO: copy over old epochs from the old engine to the new one
```
What about this TODO?

alexsporn replied:
This is imho not really needed for the MVP, this is to allow node operators to keep long history even after chain switches. The majority of nodes will follow the main chain anyway and probably never do a chain switch.
This is something that we should plan and implement as a separate feature, since it will probably mean that we need to merge the chainmanager, enginemanager and databasemanager.

jonastheis commented on packages/protocol/protocol_test.go:
```diff
@@ -639,27 +677,617 @@ func TestEngine_ShutdownResume(t *testing.T) {
 	tf2.AssertEpochState(0)
 }
+
+type NodeOnMockedNetwork struct {
```
This deserves it's own file no?

jonastheis commented on packages/protocol/protocol_test.go:
```diff
@@ -639,27 +677,617 @@ func TestEngine_ShutdownResume(t *testing.T) {
-		identity.New(identitiesMap["C"]).ID(): 25,
-		identity.New(identitiesMap["D"]).ID(): 25,
-		identity.New(identitiesMap["Z"]).ID(): 0,
+	identitiesWeights := map[ed25519.PublicKey]uint64{
+		identity.New(identitiesMap["A"]).PublicKey(): 25,
+		identity.New(identitiesMap["B"]).PublicKey(): 25,
+		identity.New(identitiesMap["C"]).PublicKey(): 25,
+		identity.New(identitiesMap["D"]).PublicKey(): 25,
+		identity.New(identitiesMap["Z"]).PublicKey(): 0,
```
Not necessarily part of this PR but this code is repeated over and over again. It should be part of some testframework to make it easy to initialize a bunch of identities with alias and weights from which the snapshot is created

alexsporn committed:
Added regions to the enginemanager.go

alexsporn committed:
Added evictionMutex to chainmanager to protect the fields that can get evicted

alexsporn committed:
Removed waits from test

alexsporn committed:
Update packages/protocol/protocol.go

Co-authored-by: Jonas Theis <4181434+jonastheis@users.noreply.github.com>

alexsporn committed:
Update packages/protocol/protocol.go

Co-authored-by: Jonas Theis <4181434+jonastheis@users.noreply.github.com>

alexsporn committed:
Removed debug log and triggering error event if cleaning up the filesystem from old engines fail

alexsporn committed:
Refactored NodeOnMockedNetwork to own package

alexsporn committed:
Added helper functions to chainmanager testframework to reduce code duplication

alexsporn committed:
Properly start and stop the scheduler

alexsporn committed:
Fixed scheduler tests

jonastheis reviewed (approved):
Good stuff

hmoog committed:
Refactor: optimized detach

hmoog commented on packages/protocol/chainmanager/commitment.go:
```diff
@@ -57,8 +58,10 @@ func (c *Commitment) Children() (children []*Commitment) {
 
-	children = make([]*Commitment, len(c.children))
-	copy(children, c.children)
+	children = make([]*Commitment, 0, len(c.children))
```
Can be replaced by `return lo.Values(c.children)`

hmoog commented on packages/protocol/chainmanager/commitment.go:
```diff
@@ -116,13 +119,43 @@ func (c *Commitment) registerChild(child *Commitment) (isSolid bool, chain *Chai
+	defer c.Unlock()
+
+	if _, has := c.children[commitment.ID()]; !has {
+		return errors.Errorf("trying to set a main child %s before registering it as a children", commitment.ID())
```
```suggestion
		return errors.Errorf("trying to set a main child %s before registering it as a child", commitment.ID())
```

hmoog commented on packages/protocol/chainmanager/fork.go:
```diff
@@ -0,0 +1,18 @@
+	ForkingPoint *commitment.Commitment
+}
+
+func (e *Fork) EndEpoch() epoch.Index {
```
Fork is a struct with public properties - having an additional getter feels unnecessary - we also use `forkedEvent.ForkingPoint.Index()` in other situations.

hmoog commented on packages/protocol/engine/tangle/booker/booker.go:
```diff
@@ -40,29 +42,33 @@ type Booker struct {
 	optsMarkerManager []options.Option[markermanager.MarkerManager[models.BlockID, *Block]]
 
-	*blockdag.BlockDAG
+	Workers *workerpool.Group
```
should be private

hmoog commented on packages/protocol/tipmanager/tipmanager.go:
```diff
@@ -228,20 +261,26 @@ func (t *TipManager) PromoteFutureTips(cm *commitment.Commitment) {
+	t.evict(index)
+}
 
+func (t *TipManager) evict(index epoch.Index) {
```
Making a utility method for a single function call with the same signature seems unnecessary just replace calls to this func to call of the wrapped func directly.

alexsporn replied:
the lowercase variant has no locking and is called from 2 places that already have the lock. The uppercase variant is public and called from the outside

alexsporn replied:
I replaced it by a call to `t.futureTips.Evict` directly

hmoog replied:
yeah thats what I meant :P

hmoog reviewed (changes requested):
LGTM apart from these small issues

alexsporn committed:
Review fixes
